tool/gravity/cli/system.go

```diff
@@ -61,164 +62,174 @@ func systemPullUpdates(env *localenv.LocalEnvironment, opsCenterURL string, runt
 		return trace.Wrap(err)
 	}
 
-	packages, err := findPackages(env.Packages, runtimePackage)
+	reqs, err := findPackages(env.Packages, runtimePackage)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	for _, req := range packagesToUpgrade(packages...) {
-		log.Debugf("Checking for update for %v.", req.filter)
+	for _, req := range reqs {
+		log := log.WithField("package", req.installedPackage)
+		log.Info("Checking for update.")
 		update, err := findPackageUpdate(env.Packages, remotePackages, req)
 		if err != nil {
 			if trace.IsNotFound(err) {
-				log.Infof("No update for: %s.", req)
+				log.Info("No update found.")
 				continue
 			}
 			return trace.Wrap(err)
 		}
-		log.Infof("Pulling update: %s.", update)
+		log.WithField("update", update).Info("Pulling update.")
 		env.Printf("Pulling update %v\n.", update)
-		pullReq := appservice.PackagePullRequest{
-			SrcPack:  remotePackages,
-			DstPack:  env.Packages,
-			Package:  update.To,
-			Progress: env.Reporter,
+		err = pullUpdate(env.Packages, remotePackages, env.Reporter, *update)
+		if err != nil {
+			return trace.Wrap(err)
 		}
-		if _, err = appservice.PullPackage(pullReq); err != nil {
-			if !trace.IsAlreadyExists(err) {
+		if update.ConfigPackage != nil {
+			err = pullUpdate(env.Packages, remotePackages, env.Reporter,
+				*update.ConfigPackage)
+			if err != nil {
 				return trace.Wrap(err)
 			}
 		}
 	}
 	return nil
 }
 
-// systemRollback rolls back system to the specified changesetID or the last update if changesetID is not specified
-func systemRollback(env *localenv.LocalEnvironment, changesetID, serviceName string, withStatus bool) (err error) {
-	changeset, err := getChangesetByID(env, changesetID)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-
-	fmt.Printf("rolling back %v\n", changeset)
+// systemUpdate searches and applies package updates if any
+func systemUpdate(env *localenv.LocalEnvironment, changesetID string, serviceName string, withStatus bool,
+	runtimePackage loc.Locator) error {
 	if serviceName != "" {
-		args := []string{"system", "rollback", "--changeset-id", changeset.ID, "--debug"}
+		args := []string{"system", "update",
+			"--changeset-id", changesetID,
+			"--runtime-package", runtimePackage.String(),
+			"--debug"}
 		if withStatus {
 			args = append(args, "--with-status")
 		}
 		return trace.Wrap(installOneshotService(env.Silent, serviceName, args))
 	}
 
-	changes := changeset.ReversedChanges()
-	rollback, err := env.Backend.CreatePackageChangeset(storage.PackageChangeset{Changes: changes})
+	reqs, err := findPackages(env.Packages, runtimePackage)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
+	var changes []storage.PackageUpdate
+	for _, req := range reqs {
```

Comment:
I'm honestly having a very hard time understanding why such big changes were needed in this file / following them. Is it to support planet config update during one of the new operations? Or is this some kind of refactoring?

Reply:
This was the effort to introduce the rotation of the planet packages yes. It is needlessly complicated and will be refactored at some point.